pkg/new-ui/v1beta1/backend.go

```diff
@@ -0,0 +1,340 @@
+package v1beta1
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"path/filepath"
+
+	"google.golang.org/grpc"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	experimentv1beta1 "github.com/kubeflow/katib/pkg/apis/controller/experiments/v1beta1"
+	api_pb_v1beta1 "github.com/kubeflow/katib/pkg/apis/manager/v1beta1"
+	"github.com/kubeflow/katib/pkg/util/v1beta1/katibclient"
+)
+
+func NewKatibUIHandler(dbManagerAddr string) *KatibUIHandler {
+	kclient, err := katibclient.NewClient(client.Options{})
+	if err != nil {
+		log.Printf("NewClient for Katib failed: %v", err)
+		panic(err)
+	}
+	return &KatibUIHandler{
+		katibClient:   kclient,
+		dbManagerAddr: dbManagerAddr,
+	}
+}
+
+// ServeIndex will return index.html for any non-API URL
+func (k *KatibUIHandler) ServeIndex(buildDir string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		fp := filepath.Join(buildDir, "static/index.html")
+		log.Printf("Sending file %s for url: %s", fp, r.URL)
+
+		// never cache index.html
+		w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate, max-age=0")
+
+		// return the contents of index.html
+		http.ServeFile(w, r, fp)
+	}
+}
+
+func (k *KatibUIHandler) connectManager() (*grpc.ClientConn, api_pb_v1beta1.DBManagerClient) {
+	conn, err := grpc.Dial(k.dbManagerAddr, grpc.WithInsecure())
+	if err != nil {
+		log.Printf("Dial to GRPC failed: %v", err)
+		return nil, nil
+	}
+	c := api_pb_v1beta1.NewDBManagerClient(conn)
+	return conn, c
+}
+
+func (k *KatibUIHandler) CreateExperiment(w http.ResponseWriter, r *http.Request) {
+	//enableCors(&w)
+	var data map[string]interface{}
+
+	json.NewDecoder(r.Body).Decode(&data)
+	dataJSON, ok := data["postData"]
+	if !ok {
+		msg := "Couldn't load the 'postData' field of the request's data"
+		log.Printf(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	jsonbody, err := json.Marshal(dataJSON)
+	if err != nil {
+		log.Printf("Marshal data for experiment failed: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	job := experimentv1beta1.Experiment{}
+	if err := json.Unmarshal(jsonbody, &job); err != nil {
+		log.Printf("Unmarshal experiment failed: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	dataMap := dataJSON.(map[string]interface{})
+	job.TypeMeta = metav1.TypeMeta{
+		APIVersion: "kubeflow.org/v1beta1",
+		Kind:       "Experiment",
+	}
+	job.ObjectMeta = metav1.ObjectMeta{
+		Name:      dataMap["metadata"].(map[string]interface{})["name"].(string),
+		Namespace: dataMap["metadata"].(map[string]interface{})["namespace"].(string),
+	}
```

Comment:
Do we need this `TypeMeta` and `ObjectMeta` filling since we add them in the submission: https://github.com/kubeflow/katib/blob/3b76c41498fce7cdbeb6cf014431d0f3fc6a4548/pkg/new-ui/v1beta1/frontend/src/app/pages/experiment-creation/experiment-creation.component.ts#L83-L88?
```suggestion
```

Reply:
I believe we do. The backed code creates an empty Experiment and manually sets the fields based on the post data.

So IIUC both of them are needed, but we should not manually set the `APIVersion` and `Kind`, for `TypeMeta` and use the values from the request [ just like we do for `ObjectMeta` ]

Reply:
@kimwnasptd Yes, but since we unmarshal the JSON data to the Experiment and we have the correct JSON Experiment in the post data, does `TypeMeta` and `ObjectMeta` is set also ?https://github.com/kubeflow/katib/blob/3b76c41498fce7cdbeb6cf014431d0f3fc6a4548/pkg/new-ui/v1beta1/backend.go#L74

Did you check how the Experiment looks like if we just unmarshal the post data ?

Reply:
Ahh I completely missed the `&job` part. Yeap we should be good removing it then.
Let me just run it and ensure we are OK and I'll push a commit

Reply:
It's as you said, we can completely omit these lines.
But I'll push a commit just for this, since after removing these lines I'll also need to remove the `metav1` import.

Reply:
Pushed the commit. I also tested and the produced `job` object has these fields set correctly.

Reply:
It's great, thank you!